param/serializer.py

```diff
@@ -0,0 +1,178 @@
+        method_name = ptype.lower()+'_' + suffix
+        return getattr(cls, method_name, None)
+
+    @classmethod
+    def parameter_schema(cls, ptype, p, safe=False):
+        if ptype in cls.unserializable_parameter_types:
+            raise UnserializableException
+        dispatch_method = cls._get_method(ptype, 'schema')
+        if dispatch_method:
+            return dispatch_method(p, safe=safe)
+        else:
+            return { "type": ptype.lower()}
+
+    @classmethod
+    def serialize(cls, ptype, value):
+        dispatch_method = cls._get_method(ptype, 'serialize')
+        if dispatch_method:
+            return dispatch_method(value)
+        else:
+            return json.dumps(value)
+
+    @classmethod
+    def deserialize(cls, ptype, string):
+        dispatch_method = cls._get_method(ptype, 'deserialize')
+        if dispatch_method:
+            return dispatch_method(string)
+        else:
+            return json.loads(string)
+
+    # Custom Serializers
+
+    @classmethod
+    def date_serialize(cls, value):
+        string = value.replace(microsecond=0).isoformat() # Test *with* microseconds.
+        return json.dumps(string)
+
+    # Custom Deserializers
+
+    @classmethod
+    def date_deserialize(cls, string):
+        return dt.datetime.fromisoformat(string)  # FIX: 3.7+ only
+
+    @classmethod
+    def tuple_deserialize(cls, string):
+        return tuple(json.loads(string))
+
+    # Custom Schemas
+
+    @classmethod
+    def date_schema(cls, p, safe=False):
+        return { "type": "string", "format": "date-time"}
+
+    @classmethod
+    def tuple_schema(cls, p, safe=False):
+        return { "type": "array"}
+
+    @classmethod
+    def number_schema(cls, p, safe=False):
+        schema = { "type": p.__class__.__name__.lower() }
+        if p.bounds is not None:
+            (low, high) = p.bounds
+            if low is not None:
+                key = 'minimum' if p.inclusive_bounds[0] else 'exclusiveMinimum'
+                schema[key] = low
+            if high is not None:
+                key = 'maximum' if p.inclusive_bounds[0] else 'exclusiveMaximum'
+                schema[key] = high
+
+        return JSONNullable(schema) if p.allow_None else schema
+
+    @classmethod
+    def integer_schema(cls, p, safe=False):
+        return cls.number_schema(p)
+
+    @classmethod
+    def numerictuple_schema(cls, p, safe=False):
+        return {"type": "array",
+                "additionalItems": { "type": "number" }}
+
+    @classmethod
+    def xycoordinates_schema(cls, p, safe=False):
+        return  {
+            "type": "array",
+            "items": [
+                {"type": "number"}, {"type": "number"}],
+            "additionalItems": False
+        }
+
+    @classmethod
+    def range_schema(cls, p, safe=False):
+        # Extend for allow None
+        return  {
+            "type": "array",
+            "items": [{"type": "number"},{"type": "number"}],
```

Comment:
This should handle the bounds on the range.

Reply:
Addressed in c353f7b